runtime/interpreter/decode.go

```diff
@@ -1877,6 +1877,70 @@ func decodeDictionaryEntries(v *DictionaryValue, content []byte) error {
 		)
 	}
 
+	// Lazily decode keys
+
+	var keysContent []byte
+	if d.isByteDecoder {
+		// Use the zero-copy method if available, for better performance.
+		keysContent, err = d.decoder.DecodeRawBytesZeroCopy()
+	} else {
+		keysContent, err = d.decoder.DecodeRawBytes()
+	}
+
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return fmt.Errorf(
+				"invalid dictionary keys encoding (@ %s): %s",
+				strings.Join(v.valuePath, "."),
+				e.ActualType.String(),
+			)
+		}
+		return err
+	}
+
+	// Lazily decode values
+
+	var valuesContent []byte
+	if d.isByteDecoder {
+		// Use the zero-copy method if available, for better performance.
+		valuesContent, err = d.decoder.DecodeRawBytesZeroCopy()
+	} else {
+		valuesContent, err = d.decoder.DecodeRawBytes()
+	}
+
+	if err != nil {
+		if e, ok := err.(*cbor.WrongTypeError); ok {
+			return fmt.Errorf(
+				"invalid dictionary values encoding (@ %s): %s",
+				strings.Join(v.valuePath, "."),
+				e.ActualType.String(),
+			)
+		}
+		return err
+	}
+
+	// Decode type
+	// TODO: store dictionary type
+	//   Option 1: convert to sema type. - Don't have the interpreter
+	//   Option 2: Store static type in array
+	_, err = d.decodeStaticType()
+	if err != nil {
```

Comment:
I think Option 2 might work, just like I commented in the PR for arrays: https://github.com/onflow/cadence/pull/1035#discussion_r660868483